base/mpfr.jl

```diff
@@ -889,22 +888,33 @@ isone(x::BigFloat) = x == Clong(1)
 @eval typemax(::Type{BigFloat}) = $(BigFloat(Inf))
 @eval typemin(::Type{BigFloat}) = $(BigFloat(-Inf))
 
-function nextfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
-          z, x, ROUNDING_MODE[])
-    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), z) != 0
-    return z
+nextfloat!(x::BigFloat) = (ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x); x)
+nextfloat(x::BigFloat) = nextfloat!(BigFloat(x, new=true))
+
+prevfloat!(x::BigFloat) = (ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x); x)
+prevfloat(x::BigFloat) = prevfloat!(BigFloat(x, new=true))
+
+function nextfloat!(x::BigFloat, n::Int)
+    n == 0 && return x
+    signbit(n) && prevfloat!(x, abs(n))
+    for i=1:n
+        ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
+    end
+    return x
 end
 
-function prevfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
-          z, x, ROUNDING_MODE[])
-    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), z) != 0
-    return z
+function prevfloat!(x::BigFloat, n::Int)
+    n == 0 && return x
+    signbit(n) && nextfloat!(x, abs(n))
+    for i=1:n
+        ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
+    end
+    return x
 end
 
+nextfloat(x::BigFloat, n::Int) = nextfloat!(BigFloat(x, new=true), n)
```

Comment:
Technically when `n == 0` we could just return `x`. Not sure it's worth the complexity though.

Reply:
Actually, the special case `n==0` is already considered in `nextfloat!()`. Note that `nextfloat()` had to return a copy of x, so anyhow we would need to call Bigfloat constructor. Thus no special case was specifically created for this function.

Reply:
> Actually, the special case `n==0` is already considered in `nextfloat!()`

But that's too late: `BigFloat(x, new=true)` has already allocated a new copy of `x` by the time `nextfloat!` gets the value.

Reply:
My argument is that, nextfloat() needs to return a new copy of x, as it does not ends with `!`. So even in case n==0, we would need to create a copy of x. Thus, no computation is wasted.

Reply:
> My argument is that, `nextfloat()` needs to return a new copy of `x`

Returning the same value is not a mutating operation. Since BigFloats are considered to be immutable it's fine to return the same value instead of allocating a new identical value.

Reply:
Does this look good?
```julia
nextfloat(x::BigFloat, n::Int) = n == 0 ? x : nextfloat!(BigFloat(x, new=true), n)
```

Reply:
Yes, that's just it.